python/ray/ray_operator/operator.py

```diff
@@ -181,162 +170,114 @@ def write_config(self) -> None:
+    cluster_identifier = (name, namespace)
+    log_prefix = ",".join(cluster_identifier)
 
-    while True:
-        # This outer loop waits for creation of a RayCluster CRD if it hasn't
-        # already been created.
-        try:
-            # Enter main event loop.
-            run_event_loop()
-        except ApiException as e:
-            if e.status == 404:
-                logger.warning("Waiting for creation of the RayCluster CRD")
-                time.sleep(5)
-            else:
-                logger.error("Failed to enter operator event loop.")
-                # Unforeseen startup error. Operator pod is
-                # likely to end up in a crash loop.
-                raise
-
-
-if __name__ == "__main__":
-    main()
+    operator_utils.check_redis_password_not_specified(
+        cluster_config, cluster_identifier)
+
+    ray_cluster = RayCluster(cluster_config)
+    ray_clusters[cluster_identifier] = ray_cluster
+    cluster_status_q.put((name, namespace, STATUS_UPDATING))
+
+    # Launch a the Ray cluster by SSHing into the pod and running
+    # the initialization commands. This will not restart the cluster
+    # unless there was a failure.
+    logger.info(f"{log_prefix}: Launching cluster.")
+    ray_cluster.create_or_update()
+    cluster_status_q.put((name, namespace, STATUS_RUNNING))
+
+
+@kopf.on.update('rayclusters')
+def update_fn(body, old, new, name, namespace, **kwargs):
+    cluster_config = operator_utils.cr_to_config(body)
+    cluster_identifier = (name, namespace)
+
+    # Check metadata.generation to determine if there's a spec change.
+    old_generation = old["metadata"]["generation"]
+    current_generation = new["metadata"]["generation"]
+    # Check metadata.labels.autoscalerRetries to see if we need to restart
+    # Ray processes.
+    old_autoscaler_retries = old.get("status", {}).get(AUTOSCALER_RETRIES_FIELD, 0)
+    autoscaler_retries = new.get("status", {}).get(AUTOSCALER_RETRIES_FIELD, 0)
+
+    # True if there's been a chamge to the spec of the custom resource,
+    # triggering an increment of metadata.generation:
+    spec_changed = current_generation > old_generation
+    # True if monitor has failed, triggering an increment of
+    # status.autoscalerRetries:
+    ray_restart_required = (autoscaler_retries >
+                            old_autoscaler_retries)
+
+    # Update if there's been a change to the spec or if we're attempting
+    # recovery from autoscaler failure.
+    if spec_changed or ray_restart_required:
+        ray_cluster = ray_clusters.get(cluster_identifier)
+        if ray_cluster is None:
+            ray_cluster = RayCluster(cluster_config)
+            ray_clusters[cluster_identifier] = ray_cluster
+
+        cluster_status_q.put((name, namespace, STATUS_UPDATING))
+
+        # Clean up the previous cluster monitor processes to prevent running multiple
+        # overlapping background threads
+        ray_cluster.clean_up_subprocess()
```

Comment:
`ray_cluster.create_or_update` does this already, so I think this line can be removed.
(https://github.com/ray-project/ray/blob/cbc24f43a5d4cf0bb533b19a4bf6ecd2ff3aba16/python/ray/ray_operator/operator.py#L118)

Moving the autoscaler to its own pod will remove the need for the weird process orchestration.

Reply:
Yes, you're right.